src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala

```diff
@@ -218,14 +218,17 @@ class LibraryService (protected val argUserInfo: UserInfo,
   }
 
   def findDocuments(criteria: LibrarySearchParams): Future[PerRequestMessage] = {
-    getEffectiveDiscoverGroups(rawlsDAO) map { userGroups =>
+    getEffectiveDiscoverGroups(rawlsDAO) flatMap { userGroups =>
       val docsFuture = searchDAO.findDocuments(criteria, userGroups)
       val idsFuture = rawlsDAO.getWorkspaces
-      val searchResults = for {
+
+      (for {
         docs <- docsFuture
         ids <- idsFuture
-      } yield updateAccess(docs, ids)
-      RequestComplete(searchResults)
+      } yield RequestComplete(updateAccess(docs, ids))) recover {
+        case ex =>
+          throw new FireCloudExceptionWithErrorReport(ErrorReport(errorMessageFromSearchException(ex)))
+      }
     }
   }
```

Comment:
The current syntax is intentional (and I've added a code comment to that effect). For-yields are, if I understand correctly, serialized. By declaring the `docsFuture` and `idsFuture` before referencing them in the for-yield, they can be parallelized. The `getEffectiveDiscoverGroups` call *needs* to come first, because its output is used as input to `findDocuments` - so it goes outside the for-yield entirely.

Reply:
I see. Thanks for the clarification!